### PR TITLE
HPCC-16870 Add method to rebuild a query from its ECL archive

### PR DESCRIPTION
**Pull request by afishbeck**
This will be really usefull when upgrading the ECL compiler, allowing
the user to rebuild their query from the exact same code.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by richardkchapman** [comment]
Typos (rebuid usefull) in commit message

**Comment by afishbeck** [comment]
@richardkchapman Fixed typos.

**Comment by richardkchapman** [comment]
There's another type (should be "its" not "it's")

**Comment by afishbeck** [comment]
Ok fixed.  That was a bad day for grammar.

File: ecl/eclcmd/queries/ecl-queries.cpp [hunk 1]
```diff
@@ -910,6 +910,273 @@ class EclCmdQueriesConfig : public EclCmdCommon
...
+    {
+        fputs("\nUsage:\n"
+            "\n"
+            "The 'queries recreate' command recompiles a query into a new workunit and republishes\n"
+            "the new workunit.  This is usefull when upgrading to a new ECL compiler and you\n"
+            "want to recompile a query with the exact same source.\n"
+            "\n"
+            "The ECL archive must be available within the workunit of the query.\n"
+            "\n"
+            "ecl recreate <target> <query> [options]\n\n"
```
Comment (richardkchapman) [inline ecl/eclcmd/queries/ecl-queries.cpp]: Should this say "ecl queries recreate" (and line below) ?

**Comment by richardkchapman** [comment]
@wangkx Can you look too please

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 2]
```diff
@@ -1645,6 +1645,201 @@ bool CWsWorkunitsEx::onWUQueryFiles(IEspContext &context, IEspWUQueryFilesReques
...
+        resp.setTimeLimit(req.getTimeLimit_isNull() ? srcQueryTree->getPropInt("@timeLimit") : req.getTimeLimit());
+        resp.setWarnTimeLimit(req.getWarnTimeLimit_isNull() ? srcQueryTree->getPropInt("@warnTimeLimit") : req.getWarnTimeLimit());
+
+        StringAttr wuid;
+        StringAttr jobname;
+
+        const char* srcQueryId = srcQueryTree->queryProp("@id");
+        const char* srcQueryName = srcQueryTree->queryProp("@name");
+
+        PROGLOG("WURecreateQuery: QuerySet %s, query %s, wuid %s", srcTarget, srcQueryId, wuid.str());
```
Comment (wangkx) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp]: Is the wuid empty?

File: esp/services/ws_workunits/ws_workunitsQuerySets.cpp [hunk 3]
```diff
@@ -1645,6 +1645,201 @@ bool CWsWorkunitsEx::onWUQueryFiles(IEspContext &context, IEspWUQueryFilesReques
...
+        waitForWorkUnitToCompile(wuid.str(), req.getWait());
+
+        Owned<IConstWorkUnit> cw(factory->openWorkUnit(wuid.str()));
+        if(!cw)
+            throw MakeStringException(ECLWATCH_CANNOT_OPEN_WORKUNIT,"Cannot open recreated workunit %s.",wuid.str());
+
+        if (jobname.length())
+        {
+            StringBuffer name;
+            if (notEmpty(jobname))
```
Comment (wangkx) [inline esp/services/ws_workunits/ws_workunitsQuerySets.cpp]: The jobname is not empty by line 1741.

**Review by wangkx — COMMENTED** [review]
@afishbeck a few comments.

**Comment by afishbeck** [comment]
@wangkx updated, please review

**Comment by wangkx** [comment]
@afishbeck looks fine.

**Pull request merged**